modules/StreamingAPI/src/main/java/org/gephi/streaming/api/GraphUpdaterEventHandler.java

```diff
@@ -0,0 +1,563 @@
+                        this.nodeRemoved(elementEvent.getElementId());
+                        break;
+                }
+                break;
+            case EDGE:
+                switch (event.getEventType()) {
+                    case ADD:
+                        EdgeAddedEvent eaEvent = (EdgeAddedEvent)event;
+                        this.edgeAdded(elementEvent.getElementId(), eaEvent.getSourceId(),
+                                eaEvent.getTargetId(), eaEvent.isDirected(),
+                                elementEvent.getAttributes());
+                        break;
+                    case CHANGE:
+                        this.edgeChanged(elementEvent.getElementId(),
+                                elementEvent.getAttributes());
+                        break;
+                    case REMOVE:
+                        this.edgeRemoved(elementEvent.getElementId());
+                        break;
+                }
+                break;
+            }
+        } else if (event instanceof FilterEvent) {
+            FilterEvent filterEvent = (FilterEvent)event;
+            applyFilter(filterEvent);
+        }
+    }
+
+    private void log(String message) {
+        if (report!=null) {
+            Issue issue = new Issue(message, Issue.Level.INFO);
+            report.logIssue(issue);
+        }
+        logger.warning(message);
+    }
+    
+    private void graphChanged(Map<String, Object> attributes) {
+        graph.writeLock();
+
+        if (attributes!=null && attributes.size() > 0) {
+            for(Map.Entry<String, Object> entry: attributes.entrySet()) {
+                graph.setAttribute(entry.getKey(), entry.getValue());
+            }
+        }
+
+        graph.writeUnlock();
+    }
+
+    private void edgeAdded(String edgeId, String fromNodeId, String toNodeId,
+            boolean directed, Map<String, Object> attributes) {
+        
+        Edge edge = graph.getEdge(edgeId);
+        if (edge!=null) {
+            log("Edge added event ignored for edge "+edgeId+": Edge already exists");
+            return;
+        }
+        
+        Node source = graph.getNode(fromNodeId);
+        if (source==null) {
+            log("Edge added event ignored for edge "+edgeId+": Source node "+fromNodeId+" not found");
+            return;
+        }
+
+        Node target = graph.getNode(toNodeId);
+        if (target==null) {
+            log("Edge added event ignored for edge "+edgeId+": Target node "+toNodeId+" not found");
+            return;
+        }
+
+        if(source!=null && target!=null) {
+            
+            edge = factory.newEdge(edgeId, source, target,0 , 1.0f, directed);
+            
+            if (attributes!=null && attributes.size() > 0) {
+                for(Map.Entry<String, Object> entry: attributes.entrySet()) {
+                    this.addEdgeAttribute(edge, entry.getKey(), entry.getValue());
+                }
+            }
+            
+            graph.writeLock();
```

Comment:
You don't need to wrap a `addEdge` call with a writeLock as it's done for you under the hood. However it could make sense to extend the write lock to the entire block. I can see an issue if the source or target node gets removed between the time you query them and the time you create and add the edge.